src/GafferScene/Cryptomatte.cpp

```diff
@@ -0,0 +1,922 @@
+}
+
+void Cryptomatte::hashChannelNames( const GafferImage::ImagePlug *parent, const Gaffer::Context *context, IECore::MurmurHash &h ) const
+{
+    FlatImageProcessor::hashChannelNames( parent, context, h );
+    inPlug()->channelNamesPlug()->hash( h );
+    outputChannelPlug()->hash( h );
+}
+
+IECore::ConstStringVectorDataPtr Cryptomatte::computeChannelNames( const Gaffer::Context *context, const GafferImage::ImagePlug *parent ) const
+{
+    StringVectorDataPtr resultData = inPlug()->channelNamesPlug()->getValue()->copy();
+    std::vector<std::string> &result = resultData->writable();
+    const std::string alphaChannel = outputChannelPlug()->getValue();
+
+    for( const auto &channelName : { "R", "G", "B" } )
+    {
+        if( find( result.begin(), result.end(), channelName ) == result.end() )
+        {
+            result.push_back( channelName );
+        }
+    }
+
+    if( alphaChannel != "" && find( result.begin(), result.end(), alphaChannel ) == result.end() )
+    {
+        result.push_back( alphaChannel );
+    }
+    
+    return resultData;
+}
+
+void Cryptomatte::hashChannelData( const GafferImage::ImagePlug *output, const Gaffer::Context *context, IECore::MurmurHash &h ) const
+{
+    FlatImageProcessor::hashChannelData( output, context, h );
+
+    inPlug()->channelDataPlug()->hash( h );
+
+    std::string cryptomatteLayer;
+    std::string alphaChannel;
+    ConstStringVectorDataPtr channelNamesData;
+    {
+        GafferImage::ImagePlug::GlobalScope globalScope( context );
+        channelNamesData = inPlug()->channelNamesPlug()->getValue();
+        cryptomatteLayer = layerPlug()->getValue();
+        alphaChannel = outputChannelPlug()->getValue();
+
+        outputChannelPlug()->hash( h );
+        layerPlug()->hash( h );
+        matteValuesPlug()->hash( h );
+        inPlug()->metadataPlug()->hash( h );
+    }
+
+    const std::string channelName = context->get<std::string>( GafferImage::ImagePlug::channelNameContextName );
+    
+    if( channelName == "R" || channelName == "G" || channelName == "B" || channelName == alphaChannel )
+    {
+        if( channelName != "R" )
+        {
+            matteChannelDataPlug()->hash( h );
+        }
+
+        GafferImage::ImagePlug::ChannelDataScope channelDataScope( context );
+        const std::string &firstDataChannel = cryptomatteLayer + g_firstDataChannelSuffix;
+        if( GafferImage::ImageAlgo::channelExists( channelNamesData->readable(), firstDataChannel ) )
+        {
+            channelDataScope.setChannelName( &firstDataChannel );
+            inPlug()->channelDataPlug()->hash( h );
+        }
+        h.append( channelName );
+    }
+}
+
+IECore::ConstFloatVectorDataPtr Cryptomatte::computeChannelData( const std::string &channelName, const Imath::V2i &tileOrigin, const Gaffer::Context *context, const GafferImage::ImagePlug *parent ) const
+{
+    const std::string alphaChannel = outputChannelPlug()->getValue();
```

Comment:
Since we're making a GlobalScope anyway, it'd be good to evaluate this inside it. This would also match the structure of `hashChannelData()` a bit more closely. Generally I find it best to match the logical structure exactly where possible, as it makes verifying the implementation a good bit easier.